NullPointerException: Cannot invoke "scala.reflect.internal.Types$Type.typeSymbol()" because the return value of "scala.reflect.internal.Trees$Tree.tpe()" is null

## reproduction steps

I can't seem to reduce the problem to a fully self-contained example (I tried for 1 hour without success), but it can be easily reproduced using a third party dependency, starting from this commit here:
https://github.com/jOOQ/jOOQ/commit/1ab673fa9a1f8fb38877883f3d44b8e08ff0d7d0

But it still also works with this one:
https://github.com/jOOQ/jOOQ/commit/5e69809283c7804373c32ae3bac2f4ff82eab599

Reproduction steps using openjdk version "17-ea" 2021-09-14 available from here https://jdk.java.net/17:

```
$ git clone https://github.com/jOOQ/jOOQ
$ cd jOOQ
$ git checkout 5e69809283c7804373c32ae3bac2f4ff82eab599
$ mkdir jOOQ-scala_2.13/src/main/scala/p
$ cat > jOOQ-scala_2.13/src/main/scala/p/S.scala
package p

import org.jooq.{Field, Record2}
import org.jooq.impl.UDTRecordImpl

abstract class S extends UDTRecordImpl[S](null) with Record2[Integer, String] {
}
$ mvn clean install -pl jooq,jooq-scala_2.13
```

## problem

The following error is produced:

```
[ERROR] ## Exception when compiling 2 sources to C:\Users\lukas\jOOQ\3.15\workspace\jOOQ-pro\jOOQ-test-codegen-scala\target\classes
java.lang.NullPointerException: Cannot invoke "scala.reflect.internal.Types$Type.typeSymbol()" because the return value of "scala.reflect.internal.Trees$Tree.tpe()" is null
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.preErase(Erasure.scala:1276)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1331)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:991)
scala.reflect.internal.Trees$Apply.transform(Trees.scala:786)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:57)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1367)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:991)
scala.reflect.internal.Trees$DefDef.$anonfun$transform$5(Trees.scala:453)
scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:43)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:38)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:32)
scala.reflect.internal.Trees$DefDef.transform(Trees.scala:451)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:57)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1357)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:991)
scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
scala.reflect.internal.Trees$Template.transform(Trees.scala:544)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$1(TypingTransformers.scala:53)
scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:43)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:38)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1367)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:991)
scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2595)
scala.reflect.internal.Trees$ClassDef.$anonfun$transform$2(Trees.scala:361)
scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:43)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:38)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:32)
scala.reflect.internal.Trees$ClassDef.transform(Trees.scala:360)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:57)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1367)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:991)
scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2622)
scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2620)
scala.reflect.internal.Trees$PackageDef.$anonfun$transform$1(Trees.scala:342)
scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:43)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:38)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:32)
scala.reflect.internal.Trees$PackageDef.transform(Trees.scala:342)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.$anonfun$transform$2(TypingTransformers.scala:55)
scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2633)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.atOwner(TypingTransformers.scala:43)
scala.tools.nsc.transform.TypingTransformers$TypingTransformer.transform(TypingTransformers.scala:38)
scala.tools.nsc.transform.Erasure$ErasureTransformer$$anon$4.transform(Erasure.scala:1367)
scala.tools.nsc.transform.Erasure$ErasureTransformer.transform(Erasure.scala:1377)
scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:183)
scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:32)
scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:454)
scala.tools.nsc.Global$GlobalPhase.run(Global.scala:401)
scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1524)
scala.tools.nsc.Global$Run.compileUnits(Global.scala:1508)
scala.tools.nsc.Global$Run.compileSources(Global.scala:1500)
scala.tools.nsc.Global$Run.compileFiles(Global.scala:1614)
xsbt.CachedCompiler0.run(CompilerBridge.scala:163)
xsbt.CachedCompiler0.run(CompilerBridge.scala:134)
xsbt.CompilerBridge.run(CompilerBridge.scala:39)
sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:92)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$7(MixedAnalyzingCompiler.scala:186)
scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:241)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4(MixedAnalyzingCompiler.scala:176)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$4$adapted(MixedAnalyzingCompiler.scala:157)
sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:157)
sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:204)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:573)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:573)
sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:174)
sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:172)
sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:459)
sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:261)
sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:414)
sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:499)
sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:401)
sbt.internal.inc.Incremental$.apply(Incremental.scala:166)
sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:573)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:491)
sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
sbt_inc.SbtIncrementalCompiler.compile(SbtIncrementalCompiler.java:218)
scala_maven.ScalaCompilerSupport.incrementalCompile(ScalaCompilerSupport.java:322)
scala_maven.ScalaCompilerSupport.compile(ScalaCompilerSupport.java:121)
scala_maven.ScalaCompilerSupport.doExecute(ScalaCompilerSupport.java:96)
scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:471)
org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.base/java.lang.reflect.Method.invoke(Method.java:568)
org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)
```

Revert jOOQ to commit `fcc258bfe82ff4e81e4148f2d9cf1bbd12c28425` and the error goes away:

```
$ git checkout fcc258bfe82ff4e81e4148f2d9cf1bbd12c28425
$ mvn clean install -pl jooq,jooq-scala_2.13
```

--- a/src/compiler/scala/tools/nsc/transform/Erasure.scala
+++ b/src/compiler/scala/tools/nsc/transform/Erasure.scala
@@ -652,7 +652,7 @@ abstract class Erasure extends InfoTransform
       val rhs = member.tpe match {
         case MethodType(Nil, FoldableConstantType(c)) => Literal(c)
         case _                                =>
-          val sel: Tree    = Select(This(root), member)
+          val sel: Tree    = gen.mkAttributedSelect(gen.mkAttributedThis(root), member)
           val bridgingCall = bridge.paramss.foldLeft(sel)((fun, vparams) => Apply(fun, vparams map Ident))
 
           maybeWrap(bridgingCall)

--- a/src/compiler/scala/tools/nsc/transform/OverridingPairs.scala
+++ b/src/compiler/scala/tools/nsc/transform/OverridingPairs.scala
@@ -53,9 +53,6 @@ abstract class OverridingPairs extends SymbolPairs {
       && !exclude(low)                 // this admits private, as one can't have a private member that matches a less-private member.
       && (lowMemberType matches (self memberType high))
     ) // TODO we don't call exclude(high), should we?
-
-    override def skipOwnerPair(lowClass: Symbol, highClass: Symbol): Boolean =
-      lowClass.isJavaDefined && highClass.isJavaDefined // javac is already checking this better than we could
   }
 
   private def bothJavaOwnedAndEitherIsField(low: Symbol, high: Symbol): Boolean = {

--- a/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
@@ -383,7 +383,7 @@ abstract class RefChecks extends Transform {
           def isOverrideAccessOK = member.isPublic || {      // member is public, definitely same or relaxed access
             (!other.isProtected || member.isProtected) &&    // if o is protected, so is m
             ((!isRootOrNone(ob) && ob.hasTransOwner(mb)) ||  // m relaxes o's access boundary
-            (other.isJavaDefined && other.isProtected))      // overriding a protected java member, see #3946 #12349
+             (other.isJavaDefined && (member.isJavaDefined || other.isProtected))) // overriding a protected java member, see #3946 #12349
           }
           if (!isOverrideAccessOK) {
             overrideAccessError()

--- a/test/files/neg/t12380.check
+++ b/test/files/neg/t12380.check
@@ -1,0 +1,8 @@
+Test.scala:1: error: incompatible type in overriding
+def m(): String (defined in trait I)
+  with def m(): Object (defined in class C);
+ found   : (): Object
+ required: (): String
+object Test extends p.J.C with p.J.I {
+       ^
+1 error

--- a/test/files/neg/t12380/J.java
+++ b/test/files/neg/t12380/J.java
@@ -1,0 +1,14 @@
+package p;
+
+public class J {
+  public static class C {
+    public Object m() { return new Object(); }
+  }
+  public interface I {
+    public String m();
+  }
+
+  public static class Test extends C implements I {
+    @Override public String m() { return ""; }
+   }
+}

--- a/test/files/neg/t12380/Test.scala
+++ b/test/files/neg/t12380/Test.scala
@@ -1,0 +1,5 @@
+object Test extends p.J.C with p.J.I {
+  def main(args: Array[String]): Unit = {
+    println((this: p.J.I).m.trim)
+  }
+}

--- a/test/files/pos/t12349b/A.java
+++ b/test/files/pos/t12349b/A.java
@@ -1,0 +1,7 @@
+package p;
+
+public class A {
+  public static class R { }
+
+  /* package-protected */ R foo() { return null; }
+}

--- a/test/files/pos/t12349b/B.java
+++ b/test/files/pos/t12349b/B.java
@@ -1,0 +1,7 @@
+package q;
+
+public class B extends p.A {
+  public static class RR extends p.A.R { }
+
+  /* package-protected */ RR foo() { return null; }
+}

--- a/test/files/pos/t12349b/Test.scala
+++ b/test/files/pos/t12349b/Test.scala
@@ -1,0 +1,1 @@
+class Test extends q.B

--- a/test/files/run/t12380/A.java
+++ b/test/files/run/t12380/A.java
@@ -1,0 +1,28 @@
+// filter: unchecked
+
+package p;
+
+public class A {
+  public static interface I {
+    public I w();
+  }
+
+  public static interface J<R extends J<R>> extends I {
+    @Override public R w();
+  }
+
+  public static interface K extends I {
+    @Override public K w();
+
+    public default String mK() { return "K"; }
+  }
+
+  /* package-private */ static class B<R extends J<R>> implements J<R> {
+    @Override public R w() { return (R) this; }
+  }
+
+  public static class C<R extends J<R>> extends B<R> implements J<R> { }
+
+  // OK in Java, also OK in Scala
+  public static class Test extends C<Test> implements K { }
+}

--- a/test/files/run/t12380/Test.scala
+++ b/test/files/run/t12380/Test.scala
@@ -1,0 +1,7 @@
+class Test extends p.A.C[Test] with p.A.K
+object Test {
+  def main(args: Array[String]): Unit = {
+    assert((new Test).w.mK == "K")
+    assert((new p.A.Test).w.mK == "K")
+  }
+}

